vm/os-windows.cpp

```diff
@@ -304,6 +304,145 @@ void handle_ctrl_c() {
+  CBT_TERMINATION_FAIL
+} CTRL_BREAK_THREAD_RESULT;
+
+static DWORD WINAPI ctrl_break_thread_proc(LPVOID parent_vm) {
+  factor_vm* vm = static_cast<factor_vm*>(parent_vm);
+  HANDLE quit_event = vm->quit_ctrl_break_event;
+
+  /* This thread will process the message queue, so we don't need a window,
+     but it is required for RegisterRawInputDevices to succeed. Any window
+     will do, so we just use one of the simpler standard window classes. */
+  HWND dummy = CreateWindow(TEXT("STATIC"), NULL, 0, 0, 0, 0, 0, HWND_MESSAGE,
+                            NULL, NULL, NULL);
+  if (dummy == NULL)
+    return CBT_CREATE_WINDOW_FAIL;
+
+  // Add HID keyboard and also ignore legacy keyboard messages.
+  RAWINPUTDEVICE Rid[1];
+  Rid[0].usUsagePage = 1; // generic desktop controls
+  Rid[0].usUsage = 6; // keyboard
+  Rid[0].dwFlags = RIDEV_NOLEGACY | RIDEV_INPUTSINK;
+  Rid[0].hwndTarget = dummy; // A hwnd is required for RIDEV_INPUTSINK.
+  if (RegisterRawInputDevices(Rid, 1, sizeof(Rid[0])) == FALSE) {
+    // Registration failed. Call GetLastError to find out more.
+    DestroyWindow(dummy);
+    return CBT_RAW_INPUT_REG_FAIL;
+  }
+
+  LPBYTE lpb = NULL;
+  UINT lpb_size = 0; // Currently allocated lpb size.
+  MSG msg;
+  msg.message = WM_NULL;
+  while (msg.message != WM_QUIT) { // Use PostQuitMessage to exit the loop.
+    if (MsgWaitForMultipleObjects(1, &quit_event, FALSE, INFINITE, QS_RAWINPUT)
+        == WAIT_OBJECT_0)
+    {
+      // quit_event is signaled, post WM_QUIT to terminate the loop.
+      PostQuitMessage(CBT_OK);
+    }
+    // A WM_INPUT or a WM_QUIT message is pending. Thread message queue must
+    // be emptied before the next MsgWaitForMultipleObjects call.
+    while (PeekMessage(&msg, NULL, WM_INPUT, WM_INPUT, PM_REMOVE)
+           && (msg.message != WM_QUIT))
+    {
+      if (msg.message == WM_INPUT) {
+        // Process the raw input message.
+        UINT need_size = 0; // Size needed for the lpb data buffer.
+        GetRawInputData((HRAWINPUT)msg.lParam, RID_INPUT, NULL, &need_size,
+                        sizeof(RAWINPUTHEADER));
+        if (need_size > lpb_size) {
+          // Grow lpb_size until it's >= need_size.
+          delete[] lpb;
+          lpb = NULL;
+          if (need_size < 0x100000) { // need_size sanity check.
+            if (lpb_size == 0)
+              lpb_size = 1;
+            // Doubling the size reduces the overall number of reallocations.
+            while (need_size > lpb_size)
+              lpb_size *= 2;
+            lpb = new(nothrow) BYTE[lpb_size];
+          }
+        }
+
+        if (lpb == NULL) {
+          // Memory allocation or need_size sanity check failed.
+          lpb_size = 0;
+          PostQuitMessage(CBT_MALLOC_FAIL);
+        } else if (GetRawInputData((HRAWINPUT)msg.lParam, RID_INPUT, lpb,
+                                   &lpb_size, sizeof(RAWINPUTHEADER))
+                   != lpb_size)
+        {
```

Comment:
I think the check should be `!= need_size`. I.e the return value of the second call to `GetRawInputData` should match the value you acquired in `need_size` in the first call.